internal/docstore/driver/codec.go

```diff
@@ -0,0 +1,701 @@
+		enc.EncodeUint(v.Uint())
+	case reflect.Float32, reflect.Float64:
+		enc.EncodeFloat(v.Float())
+	case reflect.Complex64, reflect.Complex128:
+		enc.EncodeComplex(v.Complex())
+	case reflect.String:
+		enc.EncodeString(v.String())
+	case reflect.Slice:
+		if v.IsNil() {
+			enc.EncodeNull()
+			return nil
+		}
+		fallthrough
+	case reflect.Array:
+		return encodeList(v, enc)
+	case reflect.Map:
+		return encodeMap(v, enc)
+	case reflect.Ptr:
+		if v.IsNil() {
+			enc.EncodeNull()
+			return nil
+		}
+		return encode(v.Elem(), enc)
+	case reflect.Interface:
+		if v.IsNil() {
+			enc.EncodeNull()
+			return nil
+		}
+		return encode(v.Elem(), enc)
+
+	case reflect.Struct:
+		return encodeStruct(v, enc)
+
+	default:
+		return gcerr.Newf(gcerr.InvalidArgument, nil, "cannot encode type %s", v.Type())
+	}
+	return nil
+}
+
+// Encode an array or non-nil slice.
+func encodeList(v reflect.Value, enc Encoder) error {
+	// Byte slices or byte arrays encode specially.
+	if v.Type().Elem().Kind() == reflect.Uint8 {
+		enc.EncodeBytes(v.Bytes())
+		return nil
+	}
+	n := v.Len()
+	enc2 := enc.EncodeList(n)
+	for i := 0; i < n; i++ {
+		if err := encode(v.Index(i), enc2); err != nil {
+			return err
+		}
+		enc2.ListIndex(i)
+	}
+	return nil
+}
+
+// Encode a map.
+func encodeMap(v reflect.Value, enc Encoder) error {
+	if v.IsNil() {
+		enc.EncodeNull()
+		return nil
+	}
+	keys := v.MapKeys()
+	enc2 := enc.EncodeMap(len(keys))
+	for _, k := range keys {
+		sk, err := stringifyMapKey(k)
+		if err != nil {
+			return err
+		}
+		if err := encode(v.MapIndex(k), enc2); err != nil {
+			return err
+		}
+		enc2.MapKey(sk)
+	}
+	return nil
+}
+
+// k is the key of a map. Encode it as a string.
+// Only strings, integers (signed or unsigned), and types that implement
+// encoding.TextMarshaler are supported.
+func stringifyMapKey(k reflect.Value) (string, error) {
+	// This is basically reflectWithString.resolve, from encoding/json/encode.go.
+	if k.Kind() == reflect.String {
+		return k.String(), nil
+	}
+	if tm, ok := k.Interface().(encoding.TextMarshaler); ok {
+		b, err := tm.MarshalText()
+		if err != nil {
+			return "", err
+		}
+		return string(b), nil
+	}
+	switch k.Kind() {
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return strconv.FormatInt(k.Int(), 10), nil
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return strconv.FormatUint(k.Uint(), 10), nil
+	default:
+		return "", gcerr.Newf(gcerr.InvalidArgument, nil, "cannot encode key %s of type %s", k, k.Type())
+	}
+}
+
+func encodeStruct(v reflect.Value, e Encoder) error {
+	didIt, err := e.EncodeStruct(v)
+	if didIt {
+		return err
+	}
+	fields, err := fieldCache.Fields(v.Type())
+	if err != nil {
+		return err
+	}
+	return encodeStructWithFields(v, fields, e)
+}
+
+func encodeStructWithFields(v reflect.Value, fields fields.List, e Encoder) error {
```

Comment:
Nit: not sure this is worth breaking out to a separate function....?

Reply:
It's used by `Document.Encode`, because `Document` already has the list of fields.